src/common/models/retention.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2019 Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package models
+
+import (
+	"time"
+)
+
+// TagRecord represents all pertinent metadata about a tag
+type TagRecord struct {
+	Project    *Project
+	Repository *RepoRecord
+
+	Name       string
+	CreatedAt  time.Time
+	LastPullAt time.Time
+	Labels     []*Label
+}
+
+// Filter is a tag filter in a Retention Policy Filter Chain
+type Filter interface {
+	// Process takes tags from the input channel and writes them to one of the three output channels.
+	// Tags are written to toKeep if the tags should be explicitly kept by the Filter
+	// Tags are written to toDelete if the tags should be explicitly deleted by the Filter
+	// Tags are written to next if the retention policy does not apply to the provided tag
+	//      or if the policy does not care if the tag is kept or deleted
+	//
+	// Filters do not own any of the provided channels and should **not** close them under any circumstance
+	Process(input <-chan *TagRecord, toKeep, toDelete, next chan<- *TagRecord) error
```

Comment:
I'd prefer simple interface method that merely determines what action to take for a tag, such as

```
Process(tag *TagRecord) Action
```

`Action` includes: keep, delete, ignore

Reply:
I'm all for simplifying the interface, however, I was laying this out according to the [proposal](https://github.com/goharbor/community/blob/master/proposals/5882-Tag-Retention-Policies.md).

The change you are proposing would lead to more of a classical (and probably simpler) implementation. When I was first drafting this up, channels seemed like a natural choice for flow structure.

Reply:
Thinking about this more I like the simpler interface much more than my initial design. I can re-tool here or send a PR to adjust the proposal accordingly. What do you guys think?

Reply:
Do we need to keep the proposal updated to reflect implementation details ?

Reply:
I'm fine with just changing it here. To me, it seems like more of an implementation detail and I don't think it'll affect the overall design too much.